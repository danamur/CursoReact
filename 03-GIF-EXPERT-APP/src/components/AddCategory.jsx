import { useState } from "react";
import { Button } from "react-bootstrap"

export const AddCategory = ({ onAddCategory }) => {
    
    const [inputValue, setInputValue] = useState('');

    const handleKeyPress = ({key}) => {
        if (key === "Enter" && inputValue.trim() !== '') {
            // Llama a la función onAddCategory para agregar una nueva categoría
            onAddCategory(inputValue);
            setInputValue('');
        }
    }

    const onInputChage = ({target}) => {
        setInputValue(target.value);
    }

    return (
        <>
            <div className="d-flex">
                <input 
                  type="text"
                  value={inputValue}
                  onChange={onInputChage}
                  onKeyUp={handleKeyPress}
                  placeholder="Nueva Categoria"
                  className="form-control"
                />
            </div>
        </>
    )
}
