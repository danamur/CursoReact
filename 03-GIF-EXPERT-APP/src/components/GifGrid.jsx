import { Card, Col } from "react-bootstrap"

export const GifGrid = ({ category }) => {
    return (
        <Card className="bg-dark text-light mt-4 border border border-secondary">
            <Card.Body>
                <Card.Title className="display-6 fw-bold">{category}</Card.Title>
            </Card.Body>
        </Card>
    )
}
