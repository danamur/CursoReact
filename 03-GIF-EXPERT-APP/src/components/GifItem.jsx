import { Card, Col } from 'react-bootstrap'

export const GifItem = ({ title, url }) => {
    return (
        <Col xs={12} sm={6} md={4} lg={3}>
            <div className="d-flex h-100">
                <Card className="bg-dark text-light mt-4 border border border-secondary">
                    <img src={url} alt={title} />
                    <Card.Body className="d-flex flex-column align-items-start">
                        <Card.Title className="mt-auto">
                            {title}
                        </Card.Title>
                    </Card.Body>
                </Card>
            </div>
        </Col>
    )
}
