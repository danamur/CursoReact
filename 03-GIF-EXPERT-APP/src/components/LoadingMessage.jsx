export const LoadingMessage = ({ isLoading }) => {
    return (
        isLoading && (
            <div className="d-flex justify-content-center">
                <div className="spinner-grow" style={{ width: "3rem", height: "3rem" }} role="status">
                    <span className="sr-only">Loading...</span>
                </div>
            </div>
        )
    );
};
