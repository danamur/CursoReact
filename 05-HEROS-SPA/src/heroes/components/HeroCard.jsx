import { Link } from "react-router-dom"

export const HeroCard = ({ id, name, slug, images, biography }) => {
    return (
        <>
            <div key={id} className="card animate__animated animate__zoomInRight" style={{ width: '20rem', margin: '10px' }}>
                <img src={images.lg} className="card-img-top" alt={name} />
                <div className="card-body">
                    <h5 className="card-title">{name}</h5>
                    <p className="card-text">{biography.fullName}</p>
                    <p className="card-text">
                        <small className="text-muted">{biography.firstAppearance}</small>
                    </p>
                </div>
                <div className="card-footer">
                    <Link to={`/hero/${name}`}>
                        <small className="text-body-secondary">Mas informacion</small>
                    </Link>
                </div>
            </div>
        </>
    );
};
