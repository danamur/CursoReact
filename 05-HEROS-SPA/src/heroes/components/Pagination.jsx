export const Pagination = ({ currentPage, totalPages, onPageChange }) => {
    return (
        <div className="pagination my-5 justify-content-center">
            <button
                className="btn btn-outline-secondary"
                onClick={() => onPageChange(currentPage - 1)}
                disabled={currentPage === 1}
            >
                Anterior
            </button>
            <span className="mx-3 my-auto">{currentPage}</span>
            <button
                className="btn btn-outline-secondary"
                onClick={() => onPageChange(currentPage + 1)}
            >
                Siguiente
            </button>
        </div>
    );
};
