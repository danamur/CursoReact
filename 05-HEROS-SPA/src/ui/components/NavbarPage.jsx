import { useEffect, useState } from 'react';
import { Link, NavLink, useNavigate } from 'react-router-dom';
import { Navbar, Nav, Container } from 'react-bootstrap';

export const NavbarPage = () => {

    const navigate = useNavigate();

    const onLogout = () => {
        navigate('/login', { relative: true });
    };

    const [scrolled, setScrolled] = useState(false);

    const handleScroll = () => {
        const scrollPos = window.scrollY;
        const headerHeight = document.querySelector(".navbar").offsetHeight;

        if (scrollPos >= headerHeight) {
            setScrolled(true);
        } else {
            setScrolled(false);
        }
    };

    useEffect(() => {
        window.addEventListener('scroll', handleScroll);

        // Cleanup event listener on component unmount
        return () => {
            window.removeEventListener('scroll', handleScroll);
        };
    }, []); // Empty dependency array means this effect runs once on mount

    const navbarClasses = `navbar navbar-dark bg-black fixed-top px-vw-5${scrolled ? ' scrolled shadow-sm' : ''}`;

    return (
        <nav id='navScroll' className="navbar navbar-dark bg-black fixed-top px-vw-5" tabIndex="0">
            <div className="container">
                <Link className="navbar-brand" to="/">
                    <svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" fill="currentColor" className="bi bi-stack"
                        viewBox="0 0 16 16">
                        <path
                            d="m14.12 10.163 1.715.858c.22.11.22.424 0 .534L8.267 15.34a.598.598 0 0 1-.534 0L.165 11.555a.299.299 0 0 1 0-.534l1.716-.858 5.317 2.659c.505.252 1.1.252 1.604 0l5.317-2.66zM7.733.063a.598.598 0 0 1 .534 0l7.568 3.784a.3.3 0 0 1 0 .535L8.267 8.165a.598.598 0 0 1-.534 0L.165 4.382a.299.299 0 0 1 0-.535L7.733.063z" />
                        <path
                            d="m14.12 6.576 1.715.858c.22.11.22.424 0 .534l-7.568 3.784a.598.598 0 0 1-.534 0L.165 7.968a.299.299 0 0 1 0-.534l1.716-.858 5.317 2.659c.505.252 1.1.252 1.604 0l5.317-2.659z" />
                    </svg>
                    <span className="ms-md-1 mt-1 fw-bolder me-md-5">SuperHeroes App</span>
                </Link>

                <ul className='navbar-nav mx-auto mb-2 mb-lg-0 list-group list-group-horizontal'>
                    <li className='nav-item'>
                        <NavLink className={({ isActive }) => `nav-link fs-5 ${isActive ? 'active' : ''}`} to="/marvel">
                            Marvel
                        </NavLink>
                    </li>
                    <li className='nav-item'>
                        <NavLink className={({ isActive }) => `nav-link fs-5 ${isActive ? 'active' : ''}`} to="/dc">
                            DC
                        </NavLink>
                    </li>
                    <li className='nav-item'>
                        <NavLink className={({ isActive }) => `nav-link fs-5 ${isActive ? 'active' : ''}`} to="/search">
                            Search
                        </NavLink>
                    </li>
                </ul>

                <button className='btn btn-outline-danger' onClick={onLogout}>
                    Logout
                </button>
            </div>
        </nav>
    );
};