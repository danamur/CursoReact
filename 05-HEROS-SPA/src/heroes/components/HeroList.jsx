import React, { useEffect, useState } from "react";
import { HeroCard } from "./HeroCard";
import { Pagination } from "./Pagination";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSpinner } from "@fortawesome/free-solid-svg-icons";
import { getHeroesByPublisher } from "../helpers";
import { Navigate, useLocation, useNavigate } from "react-router-dom";
import queryString from 'query-string';

const heroesPerPage = 20;

export const HeroList = ({ publisher }) => {

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    if (isNaN(q)) {
        navigate(`/`);
    }

    const [loading, setLoading] = useState(true);
    const [heroes, setHeroes] = useState([]);
    const [currentPage, setCurrentPage] = (q.length > 0) ? useState(q) : useState(1);

    useEffect(() => {
        const fetchData = async () => {
            try {
                const fetchedHeroes = getHeroesByPublisher(publisher, currentPage, heroesPerPage);
                setHeroes(fetchedHeroes);
                setLoading(false);
            } catch (error) {
                console.error("Error fetching data:", error);
                setLoading(false);
            }
        };

        fetchData();

        return () => {
            // Limpia cualquier recurso necesario al desmontar el componente
        };
    }, [publisher, currentPage]);

    const handlePageChange = (newPage) => {
        navigate(`?q=${newPage}`);

        setCurrentPage(newPage);
        setLoading(true);
    };

    return (
        <>
            {loading ? (
                <div className="text-center mt-5">
                    <h2>
                        <FontAwesomeIcon icon={faSpinner} spin className="mr-2" /> Cargando...
                    </h2>
                </div>
            ) : (
                <>
                    <div className="d-flex justify-content-between flex-wrap slide-in">
                        {heroes.map((heroe) => (
                            <HeroCard key={heroe.id} {...heroe} />
                        ))}
                    </div>

                    <Pagination
                        currentPage={currentPage}
                        totalPages={Math.ceil(heroes.length / heroesPerPage)}
                        onPageChange={handlePageChange}
                    />
                </>
            )}
        </>
    );
};