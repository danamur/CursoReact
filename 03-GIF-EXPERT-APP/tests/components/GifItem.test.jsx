import { render, screen } from "@testing-library/react";
import { GifItem } from "../../src/components/GifItem";

describe('Pruebas en <GifItem />', () => {

    const title = 'Saitama';
    const url = "http://hola.jpg/";

    test('Debería coincidir con el snapshot', () => {

        const { container } = render(<GifItem title={title} url={url} />);
        expect(container).toMatchSnapshot();

    });

    test('Debería mostrar la imagen con el URL y el ALT correctos', () => {

        render(<GifItem title={title} url={url} />);
        const { src, alt } = screen.getByRole('img');
        expect(src).toBe(url);

    });

    test('Debería mostrar el título dentro del componente', () => {

        render(<GifItem title={title} url={url} />);
        expect(screen.getByText(title)).toBeTruthy();

    });
    
});