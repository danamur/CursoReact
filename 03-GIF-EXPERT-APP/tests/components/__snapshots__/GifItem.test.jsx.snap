// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifItem /> Deber√≠a coincidir con el snapshot 1`] = `
<div>
  <div
    class="col-lg-3 col-md-4 col-sm-6 col-12"
  >
    <div
      class="d-flex h-100"
    >
      <div
        class="bg-dark text-light mt-4 border border border-secondary card"
      >
        <img
          alt="Saitama"
          src="http://hola.jpg/"
        />
        <div
          class="d-flex flex-column align-items-start card-body"
        >
          <div
            class="mt-auto card-title h5"
          >
            Saitama
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
